### ISSUE TITLE
GRU is super slow on MPS backend

### ISSUE THREAD

**Issue by hdnhan** [pytorch/pytorch#123148]
### 🐛 Describe the bug

Run the following code below, change `device` to `cpu` or `mps` to see the difference:

```python
import torch
import timeit

device = "cpu"  # cpu vs mps
gru = torch.nn.GRU(384, 256, num_layers=1, batch_first=True, bidirectional=True).to(device)


def run():
    x = torch.randn(1, 128, 384).to(device)
    gru(x)


# Warmup
for _ in range(10):
    run()

# Measure
collapsed = timeit.timeit(lambda: run(), number=1000)
print(f"Collapsed on {device}: {collapsed:.2f}s")
```

Output sample:
```bash
Collapsed on cpu: 8.87s
Collapsed on mps: 67.25s
```

I expect that computation time on MPS should be at least faster than CPU, not much slower like that. As I understand GRU is a bunch of matrix multiplications and additions. Is it a bug?

### Versions

```bash
OS: macOS 14.4.1 (arm64)
CPU: Apple M2

Python version: 3.10.14 (main, Mar 21 2024, 11:21:31) [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-14.4.1-arm64-arm-64bit

[pip3] torch==2.4.0.dev20240331
[pip3] torchaudio==2.2.0.dev20240331
[pip3] torchvision==0.19.0.dev20240331
```

**Comment by hdnhan** [comment]
If I replace `GRU` by `LSTM` like
```python
import torch
import timeit

device = "cpu"  # cpu vs mps
lstm = torch.nn.LSTM(384, 256, num_layers=1, batch_first=True, bidirectional=True).to(device)


def run():
    with torch.no_grad():
        x = torch.randn(1, 128, 384, device=device)
        lstm(x)


# Warmup
for _ in range(10):
    run()

# Measure
collapsed = timeit.timeit(lambda: run(), number=1000)
print(f"Collapsed on {device}: {collapsed:.2f}s")
```

Output sample:
```bash
Collapsed on cpu: 6.75s
Collapsed on mps: 2.37s
```

As I dive deeper, I see `LSTM` was implemented in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mps/operations/RnnOps.mm which I think it is the reason. If correct, then how can I add GRU support as I think GRU is very similar to LSTM?

**Comment by alexdremov** [comment]
> If I replace `GRU` by `LSTM` like
> 
> ```python
> import torch
> import timeit
> 
> device = "cpu"  # cpu vs mps
> lstm = torch.nn.LSTM(384, 256, num_layers=1, batch_first=True, bidirectional=True).to(device)
> 
> 
> def run():
>     with torch.no_grad():
>         x = torch.randn(1, 128, 384, device=device)
>         lstm(x)
> 
> 
> # Warmup
> for _ in range(10):
>     run()
> 
> # Measure
> collapsed = timeit.timeit(lambda: run(), number=1000)
> print(f"Collapsed on {device}: {collapsed:.2f}s")
> ```
> 
> Output sample:
> 
> ```shell
> Collapsed on cpu: 6.75s
> Collapsed on mps: 2.37s
> ```
> 
> As I dive deeper, I see `LSTM` was implemented in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mps/operations/RnnOps.mm which I think it is the reason. If correct, then how can I add GRU support as I think GRU is very similar to LSTM?

There's a long-lasting war between GRU and MPS ;)

Actually, I don't think that GRU on MPS is stable or produces valid results. Basically, GRU can be added the same way as LSTM.

But this is a rabit hole that is not as easy as I once expected.

**Comment by hdnhan** [comment]
> I don't think that GRU on MPS is stable or produces valid result

Can you evaluate this more or any GitHub issue link to this would be great?

I tried something like this, seems okay though.
```python
import torch

gru_cpu = torch.nn.GRU(384, 256, num_layers=1, batch_first=True, bidirectional=True)
gru_mps = torch.nn.GRU(384, 256, num_layers=1, batch_first=True, bidirectional=True)

gru_mps.load_state_dict(gru_cpu.state_dict())
gru_mps = gru_mps.to('mps')

x = torch.randn(1, 128, 384)
out_cpu = gru_cpu(x)[0]
out_mps = gru_mps(x.to("mps"))[0]
dif = (out_cpu - out_mps.cpu()).abs().max().item()
print("out dif:", dif)

out_cpu.sum().backward()
out_mps.sum().backward()
dif = (gru_cpu.weight_ih_l0.grad - gru_mps.weight_ih_l0.grad.cpu()).abs().max().item()
print("grad dif:", dif)
```
```bash
out dif: 9.5367431640625e-07
grad dif: 3.0517578125e-05
```

**Comment by alexdremov** [comment]
> Can you evaluate this more or any GitHub issue link to this would be great?

There are several old and recent issues that are still open like #94691 and #116601. Indeed, maybe validity was addressed in the recent updates (like yours 2.4.0).

What I still do not notice is explicit and dedicated MPS-based GRU implementation like LSTM has. So, GRU is expected to be slow.

**Comment by thegodone** [comment]
any update on this one ?

**Comment by sparkwj** [comment]
I also encountered the same problem on my Macbook M4. Gru is very slow!

**Comment by gkd720** [comment]
Yowzer! Slow for me too, on an M1Max Mac Studio:

(pytorch_env) [Gregorys-Mac-Studio]:(gkd) ~/Programs $ python GRU1.py
Collapsed on cpu: 4.90s
(pytorch_env) [Gregorys-Mac-Studio]:(gkd) ~/Programs $ vim GRU1.py
(pytorch_env) [Gregorys-Mac-Studio]:(gkd) ~/Programs $ python GRU1.py
Collapsed on mps: 94.00s

And the Activity Monitor's GPU History window had it pegged at 100%.

Anything new here? Who would do any investigation or work on this? PyTorch people? Random developers that like doing GRU models?

**Comment by guillaume-osmo** [comment]
Yes still bad performance huge margin compare to LTSM while LTSM needs more operations... So please can anyone do something (maybe you can use a LLM).
